src/hoverifier.ts

```diff
@@ -801,12 +823,134 @@ export function createHoverifier<C extends object, D, A>({
+                    hoveredTokenElement.classList.add(selectionHighlightClass)
+                }
+            })
+    )
+
+    /**
+     * For every position, emits an Observable with new values for the `documentHighlightsOrError` state.
+     * This is a higher-order Observable (Observable that emits Observables).
+     */
+    const documentHighlightObservables: Observable<Observable<{
+        eventType: SupportedMouseEvent | 'jump'
+        dom: DOMFunctions
+        target: HTMLElement
+        adjustPosition?: PositionAdjuster<C>
+        codeView: HTMLElement
+        codeViewId: symbol
+        scrollBoundaries?: HTMLElement[]
+        documentHighlightsOrError?: typeof LOADING | DocumentHighlight[] | ErrorLike | null
+        position?: HoveredToken & C
+        part?: DiffPart
+    }>> = resolvedPositions.pipe(
+        map(({ position, codeViewId, ...rest }) => {
+            if (!position) {
+                return of({
+                    documentHighlightsOrError: null,
+                    position: undefined,
+                    part: undefined,
+                    codeViewId,
+                    ...rest,
+                })
+            }
+            // Get the document highlights for that position
+            return toMaybeLoadingProviderResult(getDocumentHighlights(position)).pipe(
+                catchError((error): [MaybeLoadingResult<ErrorLike>] => [{ isLoading: false, result: asError(error) }]),
+                emitLoading<DocumentHighlight[] | ErrorLike, null>(LOADER_DELAY, null),
+                map(documentHighlightsOrError => ({
+                    ...rest,
+                    codeViewId,
+                    position,
+                    documentHighlightsOrError,
+                    part: position.part,
+                })),
+                // Do not emit anything after the code view this action came from got unhoverified
+                takeUntil(allUnhoverifies.pipe(filter(unhoverifiedCodeViewId => unhoverifiedCodeViewId === codeViewId)))
+            )
+        }),
+        share()
+    )
+
+    // Highlight the ranges returned by the document highlight provider
+    subscription.add(
+        documentHighlightObservables
+            .pipe(
+                switchMap(highlightObservable => highlightObservable),
+                switchMap(({ documentHighlightsOrError, position, adjustPosition, codeView, part, ...rest }) => {
+                    const highlights =
+                        documentHighlightsOrError &&
+                        documentHighlightsOrError !== LOADING &&
+                        !isErrorLike(documentHighlightsOrError)
+                            ? documentHighlightsOrError
+                            : []
+
+                    if (highlights.length === 0 || !position) {
+                        return of({ adjustPosition, codeView, part, ...rest, positions: of<Position[]>([]) })
+                    }
+
+                    const positions: Observable<Position>[] = []
```

Comment:
Some comments would be helpful to understand what's going on. I'd also find it a bit more readable if this was written as `combineLatest(higlights.map(...))` so it's one expression and doesn't have to mutate an array